experimental/hip/hip_device.c

```diff
@@ -0,0 +1,513 @@
+  IREE_RETURN_IF_ERROR(
+      iree_allocator_malloc(host_allocator, total_size, (void**)&device));
+  memset(device, 0, total_size);
+
+  iree_hal_resource_initialize(&iree_hal_hip_device_vtable, &device->resource);
+  iree_string_view_append_to_buffer(
+      identifier, &device->identifier,
+      (char*)device + iree_sizeof_struct(*device));
+  iree_arena_block_pool_initialize(params->arena_block_size, host_allocator,
+                                   &device->block_pool);
+  device->driver = driver;
+  iree_hal_driver_retain(device->driver);
+  device->hip_symbols = symbols;
+  device->params = *params;
+  device->hip_context = context;
+  device->hip_device = hip_device;
+  device->hip_stream = stream;
+  device->host_allocator = host_allocator;
+
+  iree_status_t status = iree_ok_status();
+
+  // Memory pool support is conditional.
+  if (iree_status_is_ok(status) && params->async_allocations) {
+    int supports_memory_pools = 0;
+    status = IREE_HIP_RESULT_TO_STATUS(
+        symbols,
+        hipDeviceGetAttribute(&supports_memory_pools,
+                              hipDeviceAttributeMemoryPoolsSupported,
+                              hip_device),
+        "hipDeviceGetAttribute");
+    device->supports_memory_pools = supports_memory_pools != 0;
+  }
+
+  // Create memory pools first so that we can share them with the allocator.
+  if (iree_status_is_ok(status) && device->supports_memory_pools) {
+    status = iree_hal_hip_memory_pools_initialize(
+        symbols, hip_device, &params->memory_pools, host_allocator,
+        &device->memory_pools);
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_allocator_create(
+        symbols, hip_device, stream,
+        device->supports_memory_pools ? &device->memory_pools : NULL,
+        host_allocator, &device->device_allocator);
+  }
+
+  if (iree_status_is_ok(status)) {
+    *out_device = (iree_hal_device_t*)device;
+  } else {
+    iree_hal_device_release((iree_hal_device_t*)device);
+  }
+  return status;
+}
+
+iree_status_t iree_hal_hip_device_create(
+    iree_hal_driver_t* driver, iree_string_view_t identifier,
+    const iree_hal_hip_device_params_t* params,
+    const iree_hal_hip_dynamic_symbols_t* symbols, hipDevice_t device,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(driver);
+  IREE_ASSERT_ARGUMENT(params);
+  IREE_ASSERT_ARGUMENT(symbols);
+  IREE_ASSERT_ARGUMENT(out_device);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_status_t status = iree_hal_hip_device_check_params(params);
+
+  // Get the main context for the device.
+  hipCtx_t context = NULL;
+  if (iree_status_is_ok(status)) {
+    status = IREE_HIP_RESULT_TO_STATUS(
+        symbols, hipDevicePrimaryCtxRetain(&context, device));
+  }
+  if (iree_status_is_ok(status)) {
+    status = IREE_HIP_RESULT_TO_STATUS(
+        symbols, hipSetDevice(IREE_HIPDEVICE_TO_DEVICE_ID(device)));
```

Comment:
From the [API](https://docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___device.html#ga43c1e7f15925eeb762195ccb5e063eae), `hipSetDevice` expects an `int` device index right? `hipDevice_t` is queried from `hipGetDevice` with a device index. Doc doesn't say `hipDevice_t` is just device index under the hood; so we cannot assume so. (Better to ask internally regarding this.) Therefore we may additionally need to keep track of the given device's index in its struct. Also, `IREE_HIPDEVICE_TO_DEVICE_ID` is not necessary here given that converts the `device` to an ID used in IREE.